src/modals/cpu-pinning/CpuPinningDataProvider.js

```diff
@@ -83,47 +111,93 @@ const getPlacementPolicyHostsIds = (vm) => {
-    },
-    cpuPinningPolicy: vm.cpu_pinning_policy,
-    cpuPinnings: getPinnings(vm),
+const mapVmToPinnedEntity = (vm) => {
+  const pinned = new PinnedEntity(vm.id, vm.name)
+
+  const sockets = +getSockets(vm)
+  const cores = +getCores(vm)
+  const threads = +getThreads(vm)
+  const cpuToPinnedCpuMap = getCpuToPinnedCpuMapForVm(vm)
+
+  if (sockets && cores && threads) {
+    for (let s = 0; s < +sockets; s++) {
+      for (let c = 0; c < +cores; c++) {
+        for (let t = 0; t < +threads; t++) {
+          const cpuId = s * cores * threads + c * threads + t
+          pinned.cpuPinningTopology.add(s, c, cpuId, cpuToPinnedCpuMap.get(cpuId))
+        }
+      }
+    }
   }
+
+  pinned.cpuCount = sockets * cores * threads
+  pinned.cpuPinningPolicy = vm.cpu_pinning_policy
+  pinned.cpuPinningString = getCpuPinningString(vm)
+  return pinned
 }
 
-const mapHostToModel = (host) => {
-  return {
-    id: host.id,
-    name: host.name,
-    cpus: getCpus(host),
+const createCpuPinningTopology = (cpuUnits) => {
+  const cpuPinningTopology = new Topology()
+
+  cpuUnits?.host_cpu_unit?.forEach(cpuUnit => {
+    const socketId = +cpuUnit.socket_id
+    const coreId = +cpuUnit.core_id
+    const cpuId = +cpuUnit.cpu_id
+    const exclusivelyPinned = cpuUnit.vms?.vm.some(vm => CpuPinningPolicy.isExclusive(vm.cpu_pinning_policy))
+    const pinnedEntities = (cpuUnit.vms?.vm.map(vm => vm.name) || []).sort()
+    if (cpuUnit.runs_vdsm) {
+      pinnedEntities.unshift('VDSM')
+    }
+
+    cpuPinningTopology.add(socketId, coreId, cpuId, pinnedEntities, exclusivelyPinned)
+  })
+
+  return cpuPinningTopology
+}
+
+const mapHostToPinnedEntity = (host, cpuUnits) => {
+  const pinned = new PinnedEntity(host.id, host.name)
+
+  if (host.cpu?.topology) {
+    const topology = host.cpu.topology
+    pinned.cpuCount = topology.sockets * topology.cores * topology.threads
   }
+
+  pinned.cpuPinningTopology = createCpuPinningTopology(cpuUnits)
+
+  return pinned
 }
 
-const mapHostsToModel = (hosts) => {
-  return hosts.map((host) => mapHostToModel(host))
+const mapHostsToPinnedEntities = (hosts) => {
+  return hosts.map((host) => mapHostToPinnedEntity(host))
 }
 
-const CpuPinningDataProvider = ({ children, vmId }) => {
+const CpuPinningDataProvider = ({ children, vmId, hostId }) => {
   const fetchData = async () => {
+    if (vmId) {
+      return fetchDataVmView()
+    } else if (hostId) {
+      return fetchDataHostView()
+    } else {
+      throw new Error('At least one of vmId, hostId needs to be specified')
+    }
+  }
+
+  const fetchDataVmView = async () => {
     const vm = await fetchVm(vmId)
     const hosts = await fetchHosts(getHostIds(vm))
     return {
-      vm: mapVmToModel(vm),
-      hosts: mapHostsToModel(hosts),
+      mainEntity: mapVmToPinnedEntity(vm),
+      pinnedEntities: mapHostsToPinnedEntities(hosts),
+    }
+  }
+
+  const fetchDataHostView = async () => {
+    const host = await fetchHost(hostId)
```

Comment:
The follow is not implemented at the moment (and won't be until beta freeze). How would I execute them in parallel? Promise.all() would do that?

Reply:
I don't see why `follow` wouldn't work - sounds like a good idea

Reply:
I think it needs to be implemented on the backend.

Reply:
When I try: `http://localhost.localdomain:8080/ovirt-engine/api/hosts/a96595e2-00ee-4bd4-b8d8-f219dfe78a20/?follow=cpuunits`
I get:
```
	at org.jboss.resteasy.resteasy-jaxrs@3.15.1.Final//org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:141)
	at org.jboss.resteasy.resteasy-jaxrs@3.15.1.Final//org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:104)
	at org.jboss.resteasy.resteasy-jaxrs@3.15.1.Final//org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:440)
	... 100 more
Caused by: java.lang.IllegalStateException: Problem following link: /ovirt-engine/api/hosts/a96595e2-00ee-4bd4-b8d8-f219dfe78a20/cpuunits
	at org.ovirt.engine.api.restapi-jaxrs//org.ovirt.engine.api.restapi.resource.utils.LinkFollower.fetch(LinkFollower.java:265)
	at org.ovirt.engine.api.restapi-jaxrs//org.ovirt.engine.api.restapi.resource.utils.LinkFollower.fetchData(LinkFollower.java:210)
	... 121 more
Caused by: java.lang.NoSuchMethodException: org.ovirt.engine.api.restapi.resource.BackendHostResource.getCpuUnitsResource(java.lang.String)
	at java.base/java.lang.Class.getMethod(Class.java:2108)
	at org.ovirt.engine.api.restapi-jaxrs//org.ovirt.engine.api.restapi.resource.ResourceLocator.locateResource(ResourceLocator.java:46)
	at org.ovirt.engine.api.restapi-jaxrs//org.ovirt.engine.api.restapi.resource.utils.LinkFollower.fetch(LinkFollower.java:256)
	... 122 more
```

Reply:
The LinkFollower complains about missing ` HostCpuUnitResource getCpuUnitsResource(String id)` method.
The method is missing because the model is incomplete - we have the resource part only for the collection but not the singular resource: 
1. HostCpuUnits ->  HostCpuUnitsResource -> BackendHostCpuUnitsResource
2. HostCpuUnit -> ? -> ?

Reply:
> The follow is not implemented at the moment (and won't be until beta freeze). How would I execute them in parallel? Promise.all() would do that?

Fetching in parallel inside an async function follows the form:
```js
  const [ resultA, resultB, resultC ] = await Promise.all([
    fetchA(),
    fetchB(),
    fetchC(),
  ])
```